aten/src/THC/THCCachingAllocator.cpp

```diff
@@ -148,28 +167,23 @@ struct THCCachingAllocator
   }
 
   /** allocates a block which is safe to use from the provided stream */
-  cudaError_t malloc(void** devPtr, size_t size, cudaStream_t stream)
+  void malloc(void** devPtr, size_t size, cudaStream_t stream)
   {
     std::lock_guard<std::mutex> lock(mutex);
 
     int device;
-    cudaError_t err = cudaGetDevice(&device);
-    if (err != cudaSuccess) {
-      return err;
-    }
+    AT_CUDA_CHECK(cudaGetDevice(&device));
 
-    err = process_events();
-    if (err != cudaSuccess) {
-      return err;
-    }
+    // process outstanding cudaEvents
+    process_events();
 
     size = round_size(size);
     bool small = size <= kSmallAlloc;
 
     DeviceStats &stats = get_stats_for_device(device);
 
     Block search_key(device, stream, size);
-    auto& free_blocks = small ? large_blocks : small_blocks;
+    auto& free_blocks = small ? small_blocks : large_blocks;
```

Comment:
Could this have been the culprit?

Reply:
If small allocations are routed to large blocks, they can't be freed anymore and will cause more fragmentation then if they are always routed to the small blocks.

Reply:
No. The names were swapped, but the logic was consistent.

Reply:
Hmmm how is it possible that this is equivalent? I can’t see where’s the other swap that would cancel this one out.

Reply:
In `free_block()`. Other than insertion and removal, the `small_blocks` and `large_blocks` block lists are treated identically.

(Small allocations are rounded differently from large allocations, but that's dependent on the variable `small` not the choice of block list)